src/mlpack/methods/ann/rbm_impl.hpp

```diff
@@ -0,0 +1,236 @@
+/**
+ * @file rbm_impl.hpp
+ * @author Kris Singh
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_ANN_RBM_IMPL_HPP
+#define MLPACK_METHODS_ANN_RBM_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "rbm.hpp"
+
+#include <mlpack/core.hpp>
+
+#include <mlpack/methods/ann/activation_functions/softplus_function.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial neural networks. */ {
+
+
+template<typename InitializationRuleType, typename RBMPolicy>
+RBM<InitializationRuleType, RBMPolicy>::RBM(
+    arma::Mat<ElemType> predictors,
+    InitializationRuleType initializeRule,
+    RBMPolicy rbmPolicy,
+    const size_t numSteps,
+    const size_t negSteps,
+    const bool useMonitoringCost,
+    const bool persistence):
+    rbmPolicy(rbmPolicy),
+    initializeRule(initializeRule),
+    numSteps(numSteps),
+    negSteps(negSteps),
+    useMonitoringCost(useMonitoringCost),
+    persistence(persistence),
+    reset(false)
+{
+  numFunctions = predictors.n_cols;
+  this->predictors = std::move(predictors);
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+void RBM<InitializationRuleType, RBMPolicy>::Reset()
+{
+  size_t weight = rbmPolicy.Parameters().n_elem;
+
+  positiveGradient.set_size(weight, 1);
+  negativeGradient.set_size(weight, 1);
+  negativeSamples.set_size(rbmPolicy.VisibleSize(), 1);
+  tempNegativeGradient.set_size(weight, 1);
+
+  parameter.zeros(weight, 1);
+  positiveGradient.zeros();
+  negativeGradient.zeros();
+  tempNegativeGradient.zeros();
+  initializeRule.Initialize(parameter, parameter.n_elem, 1);
+
+  rbmPolicy.Parameters() = arma::Mat<ElemType>(parameter.memptr(), weight, 1,
+      false, false);
+  rbmPolicy.Reset();
+
+  reset = true;
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+template<typename OptimizerType>
+void RBM<InitializationRuleType, RBMPolicy>::Train(
+    const arma::Mat<ElemType>& predictors,
+    OptimizerType& optimizer)
+{
+  numFunctions = predictors.n_cols;
+  this->predictors = std::move(predictors);
+
+  if (!reset)
+    Reset();
+
+  // Train the model.
+  Timer::Start("rbm_optimization");
+  optimizer.Optimize(*this, parameter);
+  Timer::Stop("rbm_optimization");
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+double RBM<InitializationRuleType, RBMPolicy>::
+    FreeEnergy(arma::Mat<ElemType>&& input)
+{
+  return rbmPolicy.FreeEnergy(std::move(input));
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+double RBM<InitializationRuleType, RBMPolicy>::Evaluate(
+    const arma::Mat<ElemType>& /* parameters*/,
+    const size_t i,
+    const size_t batchSize)
```

Comment:
@lozhnikov This argument is currently unused, can anything be done here? Same with [here](https://github.com/mlpack/mlpack/pull/1208/files#diff-4ba957d6d102420d5311acd6af750ca8R174)

Reply:
Also, should this be updated in `..ann/rbm/` subdirectory files as well?

Reply:
Actually, it depends on the optimizers. You can just comment out the name of the argument e.g.
```
const size_t i,
const size_t /*batchSize*/)
```
in order to avoid the warning.

Reply:
@lozhnikov I've updated the changes here. The `BinaryRBM` Classification test is working well now, but seemingly the `ssRBM` test has taken a hit (went down from an `80.6` to `75.2`, a little below the intended baseline). Can you please check if the implementation below is correct?